pkg/apis/sessions/session_state.go

```diff
@@ -59,78 +64,70 @@ func (s *SessionState) String() string {
-		ss.Email = s.Email
-		ss.User = s.User
-		ss.PreferredUsername = s.PreferredUsername
-	} else {
-		ss = *s
-		for _, s := range []*string{
-			&ss.Email,
-			&ss.User,
-			&ss.PreferredUsername,
-			&ss.AccessToken,
-			&ss.IDToken,
-			&ss.RefreshToken,
-		} {
-			err := into(s, c.Encrypt)
-			if err != nil {
-				return "", err
-			}
+// EncodeSessionState returns an encrypted, lz4 compressed, MessagePack encoded session
+func (s *SessionState) EncodeSessionState(c encryption.Cipher, compress bool) ([]byte, error) {
+	packed, err := msgpack.Marshal(s)
+	if err != nil {
+		return nil, err
+	}
+
+	if !compress {
+		return c.Encrypt(packed)
+	}
+
+	compressed, err := lz4Compress(packed)
+	if err != nil {
+		return nil, err
+	}
+	return c.Encrypt(compressed)
+}
+
+// DecodeSessionState decodes a LZ4 compressed MessagePack into a Session State
+func DecodeSessionState(data []byte, c encryption.Cipher, compressed bool) (*SessionState, error) {
+	decrypted, err := c.Decrypt(data)
+	if err != nil {
+		return nil, err
+	}
+
+	packed := decrypted
+	if compressed {
+		packed, err = lz4Decompress(decrypted)
+		if err != nil {
+			return nil, err
 		}
 	}
 
-	b, err := json.Marshal(ss)
-	return string(b), err
+	var ss *SessionState
+	err = msgpack.Unmarshal(packed, &ss)
+	if err != nil {
+		return nil, err
+	}
+
+	return ss, nil
 }
 
-// DecodeSessionState decodes the session cookie string into a SessionState
-func DecodeSessionState(v string, c encryption.Cipher) (*SessionState, error) {
+// LegacyV5DecodeSessionState decodes a legacy JSON session cookie string into a SessionState
+func LegacyV5DecodeSessionState(v string, c encryption.Cipher) (*SessionState, error) {
 	var ss SessionState
 	err := json.Unmarshal([]byte(v), &ss)
 	if err != nil {
 		return nil, fmt.Errorf("error unmarshalling session: %w", err)
 	}
 
-	if c == nil {
-		// Load only Email and User when cipher is unavailable
-		ss = SessionState{
-			Email:             ss.Email,
-			User:              ss.User,
-			PreferredUsername: ss.PreferredUsername,
-		}
-	} else {
-		// Backward compatibility with using unencrypted Email or User
-		// Decryption errors will leave original string
-		err = into(&ss.Email, c.Decrypt)
-		if err == nil {
-			if !utf8.ValidString(ss.Email) {
-				return nil, errors.New("invalid value for decrypted email")
-			}
-		}
-		err = into(&ss.User, c.Decrypt)
-		if err == nil {
-			if !utf8.ValidString(ss.User) {
-				return nil, errors.New("invalid value for decrypted user")
-			}
-		}
-
-		for _, s := range []*string{
-			&ss.PreferredUsername,
-			&ss.AccessToken,
-			&ss.IDToken,
-			&ss.RefreshToken,
-		} {
-			err := into(s, c.Decrypt)
-			if err != nil {
-				return nil, err
-			}
+	for _, s := range []*string{
+		&ss.User,
+		&ss.Email,
+		&ss.PreferredUsername,
+		&ss.AccessToken,
+		&ss.IDToken,
+		&ss.RefreshToken,
+	} {
+		err := into(s, c.Decrypt)
+		if err != nil || !utf8.ValidString(*s) {
+			return nil, err
```

Comment:
If the decryption succeeded but the string isn't valid UTF8, err is nil here so we return `nil, nil`, I don't think that's the intention, perhaps this should be split into two if statements?